### PR TITLE
[move core modules] add logic for burning Libra

### PR DESCRIPTION
**Pull request by sblackshear**
Previously, only Libra with value zero could be burned. This moves us a bit closer to reality by introducing a flow for burning:
    
- Funds to be burned are first moved to a staging area (a `Preburn` resource under the sender's account).
- Funds will wait in the staging area while offline operations to sell the underlying assets take place.
- Once the underlying assets have been sold, the new burn operation destroys funds that are held in the `Preburn` resource.
- The `Preburn` resource supports concurrent burn requests. In the case that multiple requests are active, the oldest one will be fulfilled first.
- Only holders of `Preburn` resources have the authority to initiate burn requests. Only the holder of the `MintCapability` has the authority to create new `Preburn` resources.

## Motivation

Working toward a more realistic burning flow.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

New E2E tests

### CODE REVIEW

File: language/stdlib/modules/libra_coin.move [hunk 1]
```diff
@@ -15,20 +15,33 @@ module LibraCoin {
...
     }
 
-    // Return a reference to the MintCapability published under the sender's account. Fails if the
-    // sender does not have a MintCapability.
-    // Since only the Association account has a mint capability, this will only succeed if it is
-    // invoked by a transaction sent by that account.
-    public fun mint_with_default_capability(amount: u64): T acquires MintCapability, MarketCap {
+    // Return `amount` coins created using the MintCapability published under the sender's account.
+    // Fails if the sender does not have a published MintCapability.
+    public fun mint_with_capability(amount: u64): T acquires MintCapability, MarketCap {
```
Comment (bob-wilson) [inline language/stdlib/modules/libra_coin.move]: This "mint_with_capability" name seems wrong to me. Compare to some other functions:

- deposit_with_metadata: similar to deposit() but with an extra metadata argument
- deposit_with_sender_and_metadata: similar to deposit_with_metadata() but with the sender argument as well
- withdraw_with_capability: similar to other withdraw functions but with an explicit WithdrawalCapability argument
- rotate_authentication_key_with_capability: similar to rotate_authentication_key but with an explicit KeyRotationCapability argument

For consistency, we should rename mint() to mint_with_capability and then mint_with_default_capability() can be just plain "mint".

File: language/stdlib/modules/libra_coin.move [hunk 2]
```diff
@@ -15,20 +15,33 @@ module LibraCoin {
...
         mint(amount, borrow_global<MintCapability>(Transaction::sender()))
     }
 
+    // Move `coin` to the preburn holding area using the MintCapability published under the sender'
+    // account. Fails if the sender does not have a published MintCapability.
+    public fun preburn_with_capability(coin: T) acquires MintCapability, MarketCap {
```
Comment (bob-wilson) [inline language/stdlib/modules/libra_coin.move]: Ditto for the "preburn" and "burn" functions here.... I don't think they should have "_with_capability" in their names.

File: language/stdlib/modules/libra_coin.move [hunk 3]
```diff
@@ -85,6 +123,14 @@ module LibraCoin {
         T{ value: amount }
     }
 
+    // Return a coin new coin whose value is coin_ref.value
```
Comment (bob-wilson) [inline language/stdlib/modules/libra_coin.move]: Typo: should be "...a new coin...", I think.

File: language/stdlib/modules/libra_coin.move (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_coin.move]: Good calls, will fix.

File: language/stdlib/modules/libra_coin.move (hunk 3 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_coin.move]: Agreed, but I actually don't end up needing this procedure in the revised module.

File: language/ir-testsuite/tests/payments/minting.mvir [hunk 4]
```diff
@@ -1,16 +0,0 @@
-import 0x0.LibraAccount;
```
Comment (sblackshear) [inline language/ir-testsuite/tests/payments/minting.mvir]: I migrated this to the source language tests.

**Comment by dmitri-perelman** [comment]
This is awesome! Will let the Bob and Tim do the actual code review, but the high level approach is neat :)

File: language/stdlib/modules/libra_coin.move [hunk 5]
```diff
@@ -15,30 +16,124 @@ module LibraCoin {
...
 
-    // Return a reference to the MintCapability published under the sender's account. Fails if the
-    // sender does not have a MintCapability.
-    // Since only the Association account has a mint capability, this will only succeed if it is
-    // invoked by a transaction sent by that account.
-    public fun mint_with_default_capability(amount: u64): T acquires MintCapability, MarketCap {
-        mint(amount, borrow_global<MintCapability>(Transaction::sender()))
+    // A holding area where funds that will subsequently be burned wait while their underyling
+    // assets are sold off-chain. Multiple preburn requests can be in progress at the same time,
+    // but they must be resolved in LIFO order.
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/libra_coin.move]: yes #soashamed

**Pull request closed without merging**